Edit Product: inventory settings UI

jaclync opened the pull request:
UI part for #1424 

## Summary

This PR implements the UI for the Product Inventory Settings without the remote update functionality (the next task).

## Changes

- Created a new enum `ProductBackordersSetting` based on `Product`'s `backordersKey` as the raw value. Added a calculated property `backordersSetting: ProductBackordersSetting` for `Product`
- Added `StringInputFormatter: UnitInputFormatter` for temporary SKU input view model
- Created 2 functions for `Product` extension to create view models for SKU and stock quantity
- Created `ProductBackordersSettingListSelectorDataSource` for backorders setting list selector + basic unit tests
- Created `ProductInventorySettingsViewController` with a table view that shows 3 sections of inventory settings. Implemented cell did select actions
- Added `keyboardType: UIKeyboardType` to `UnitInputViewModel` and hid the unit label if empty

## Testing

Prerequisite: at least 2 Products, one with Manage Stock enabled and the other disabled

- Launch the app
- Go to the Products tab
- Tap on a Product
- Tap "Edit" in the navigation bar
- Tap the inventory settings row --> should see 3 sections of data: SKU, Manage Stock and its corresponding fields when it's enabled/disabled, and whether a Product can only be sold once per order
- Toggle the "Manage Stock" switch --> the fields should change
- Repeat the above for Products with Manage Stock enabled and disabled

## Example screenshots

@kyleaparker I'm not sure about the navigation bar title of the Backorders allowed selector screen, I used what's on wp-admin for the first pass

manage stock disabled | manage stock enabled | stock status selector | backorders selector
-- | -- | -- | --
![Simulator Screen Shot - iPhone 8 - 2019-12-18 at 17 59 40](https://user-images.githubusercontent.com/1945542/71076529-d89fa500-21c0-11ea-8643-77f08611355a.png) | ![Simulator Screen Shot - iPhone 8 - 2019-12-18 at 18 05 19](https://user-images.githubusercontent.com/1945542/71076611-fa992780-21c0-11ea-981f-1ed0f2b8f342.png) | ![Simulator Screen Shot - iPhone 8 - 2019-12-18 at 18 00 10](https://user-images.githubusercontent.com/1945542/71076531-d9383b80-21c0-11ea-8b02-4cc5a9e856be.png) | ![Simulator Screen Shot - iPhone 8 - 2019-12-18 at 18 03 13](https://user-images.githubusercontent.com/1945542/71076532-d9383b80-21c0-11ea-812e-55f7631117a3.png)

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mindgraffiti commented on Networking/Networking/Model/Product/ProductBackordersSetting.swift:
```diff
@@ -0,0 +1,66 @@
+    case allowed
+    case allowedAndNotifyCustomer
+    case notAllowed
+    case custom(String) // in case there are extensions modifying backorders setting
```
yay!

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Inventory Settings/ProductInventorySettingsViewController.swift:
```diff
@@ -0,0 +1,295 @@
+    private var backordersSetting: ProductBackordersSetting?
+
+    // Editable data - manage stock disabled.
+    private var stockStatus: ProductStockStatus?
```
nit: comment formatting `///`

jaclync replied:
Generally, I use `///` for property/function/struct/class level comments and `//` for comments on a block of properties or misc notes. Lemme know if there are some official/unofficial docs for best practices!

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Inventory Settings/ProductInventorySettingsViewController.swift:
```diff
@@ -0,0 +1,295 @@
+            stockSection = Section(rows: [.manageStock, .stockQuantity, .backorders])
+        } else {
+            stockSection = Section(rows: [.manageStock, .stockStatus])
+        }
```
This could be condensed to:

```
var stockSection = Section(rows: [.manageStock, .stockStatus])
if (manageStockEnabled) {
    stockSection.append(.backorders)
}
```

jaclync replied:
I slightly prefer using `let` to `var` so that it can only be set in the same conditional context, otherwise it could be set to a different value later in the function

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Inventory Settings/ProductInventorySettingsViewController.swift:
```diff
@@ -0,0 +1,295 @@
+// MARK: - Input changes handling
+//
+private extension ProductInventorySettingsViewController {
+    func handleSKUChange(sku: String?) {
```
```suggestion
    func handleSKUChange(_ sku: String?) {
```

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Inventory Settings/ProductInventorySettingsViewController.swift:
```diff
@@ -0,0 +1,295 @@
+        self.sku = sku
+    }
+
+    func handleStockQuantityChange(stockQuantity: String?) {
```
```suggestion
    func handleStockQuantityChange(_ stockQuantity: String?) {
```

jaclync commented:
updated the PR based on the PR feedback, and removed the back bar button text in https://github.com/woocommerce/woocommerce-ios/pull/1641/commits/5a9f1c80f167f19f651221339c0a6527ce779fe8
